src/MultipartParser.php

```diff
@@ -0,0 +1,193 @@
+ * @license http://www.gnu.org/licenses/gpl.html GPL License 3
+ */
+class MultipartParser
+{
+    /**
+     * @var string
+     */
+    protected $input;
+
+    /**
+     * @var string
+     */
+    protected $boundary;
+
+    /**
+     * Contains the resolved posts
+     *
+     * @var array
+     */
+    protected $post = [];
+
+    /**
+     * Contains the resolved files
+     *
+     * @var array
+     */
+    protected $files = [];
+
+    /**
+     * @param $input
+     * @param $boundary
+     */
+    public function __construct($input, $boundary)
+    {
+        $this->input = $input;
+        $this->boundary = $boundary;
+    }
+
+    /**
+     * @return array
+     */
+    public function getPost()
+    {
+        return $this->post;
+    }
+
+    /**
+     * @return array
+     */
+    public function getFiles()
+    {
+        return $this->files;
+    }
+
+    /**
+     * Do the actual parsing
+     */
+    public function parse()
+    {
+        $blocks = $this->split($this->boundary);
+
+        foreach ($blocks as $value) {
+            if (empty($value)) {
+                continue;
+            }
+
+            $this->parseBlock($value);
+        }
+    }
+
+    /**
+     * @param $boundary string
+     * @returns Array
+     */
+    protected function split($boundary)
+    {
+        $boundary = preg_quote($boundary);
+        $result = preg_split("/\\-+$boundary/", $this->input);
+        array_pop($result);
+        return $result;
+    }
+
+    /**
+     * Decide if we handle a file, post value or octet stream
+     *
+     * @param $string string
+     * @returns void
+     */
+    protected function parseBlock($string)
+    {
+        if (strpos($string, 'filename') !== false) {
+            $this->file($string);
+            return;
+        }
+
+        // This may never be called, if an octet stream
+        // has a filename it is catched by the previous
+        // condition already.
+        if (strpos($string, 'application/octet-stream') !== false) {
+            $this->octetStream($string);
+            return;
+        }
+
+        $this->post($string);
+    }
+
+    /**
+     * Parse a raw octet stream
+     *
+     * @param $string
+     * @return array
+     */
+    protected function octetStream($string)
+    {
+        preg_match('/name=\"([^\"]*)\".*stream[\n|\r]+([^\n\r].*)?$/s', $string, $match);
+
+        $this->addResolved('post', $match[1], $match[2]);
+    }
+
+    /**
+     * Parse a file
+     *
+     * @param $string
+     * @return array
+     */
+    protected function file($string)
+    {
+        preg_match('/name=\"([^\"]*)\"; filename=\"([^\"]*)\"[\n|\r]+([^\n\r].*)?\r$/s', $string, $match);
+        preg_match('/Content-Type: (.*)?/', $match[3], $mime);
+
+        $content = preg_replace('/Content-Type: (.*)[^\n\r]/', '', $match[3]);
+        $content = ltrim($content, "\r\n");
+
+        $path = tempnam(sys_get_temp_dir(), "php");
+        $err = file_put_contents($path, $content);
```

Comment:
A stream per fileupload. Ping me on IRC or on twitter if you have questions about that. Happy to help :).